### ISSUE TITLE
Java-1.8.eb modulerc file fails to install with EasyBuild-4.2.1 due to changes in modules.py

### ISSUE THREAD

**Issue by casparvl** [easybuilders/easybuild-framework#3376]
**Issue**

With EasyBuild-4.2.0 `eb Java-1.8.eb -f` worked without problems, but with EasyBuild-4.2.1 I get:

```
== 2020-07-01 13:12:01,272 modules.py:596 DEBUG Module Java/1.8.0_192 not found via module avail/show, checking whether it is a wrapper
== 2020-07-01 13:12:01,273 modules.py:502 DEBUG Found /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/modulefiles/all/Java/.modulerc that may define Java/1.8.0_192 as a wrapper for a module file
== 2020-07-01 13:12:01,273 modules.py:502 DEBUG Found /sw/arch/RedHatEnterpriseServer7/EB_production/2019/modulefiles/lang/Java/.modulerc that may define Java/1.8.0_192 as a wrapper for a module file
== 2020-07-01 13:12:01,274 modules.py:603 DEBUG Result for existence check of Java/1.8.0_192 module: False
== 2020-07-01 13:12:01,382 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/base/exceptions.py:124 in __init__): Can't get value from a non-existing module Java/1.8.0_192 (at easybuild/tools/modules.py:691 in get_value_from_modulefile)
```

I have pinpointed this to the `modules.py`, because if I copy the `modules.py` from 4.2.0 into my 4.2.1 installation, my EasyBuild 4.2.1 can suddenly complete `eb Java-1.8.eb -f` succesfully.

Quite a number of things changed in the `modules.py` with the last update, but most of it is logging. I suspect the change in `mod_exists_via_show(mod_name)` to be the breaking change. I will proceed to modify the modules.py to only roll back that change.

**Comment by casparvl** [comment]
I changed the `mod_exists_via_show` function of EasyBuild-4.2.1 to:

```
def mod_exists_via_show(mod_name):
            """
            Helper function to check whether specified module name exists through 'module show'.

            :param mod_name: module name
            """

            # Old mod exists:
            mod_exists_regex_template_local = r'^\s*\S*/%s.*:\s*$'
            mod_exists_regex = mod_exists_regex_template_local % re.escape(mod_name)
            txt = self.show(mod_name)
            old_res = bool(re.search(mod_exists_regex, txt, re.M))
            self.log.debug("Old mod_exists_via_show would return %s" % old_res)

            stderr = self.show(mod_name)
            res = False
            # Parse the output:
            # - Skip whitespace
            # - Any error -> Module does not exist
            # - Check first non-whitespace line for something that looks like an absolute path terminated by a colon
            mod_exists_regex = r'\s*/.+:\s*'
            for line in stderr.split('\n'):
                if OUTPUT_MATCHES['whitespace'].search(line):
                    continue
                if OUTPUT_MATCHES['error'].search(line):
                    break
                if re.match(mod_exists_regex, line):
                    res = True
                break
            self.log.debug("New mod_exists_via_show will return %s" % res)
            return res
```
I.e. it first goes through the logic that was in EasyBuild 4.2.0 and prints whether the function would return true or false. Then, it goes through the EasyBuild v 4.2.1 logic, prints whether that would return true or false, and return that.

My (debugging) output now shows:

```
== 2020-07-01 13:37:29,132 modules.py:556 DEBUG Old mod_exists_via_show would return True
...
== 2020-07-01 13:37:29,133 modules.py:573 DEBUG New mod_exists_via_show will return False
```
Finally, changing the function further by returning `old_res` instead of `res`, the build now completes.

In other words: this piece of code is indeed the breaking change.

**Comment by casparvl** [comment]
Running the original EasyBuild 4.2.1 `modules.py` with some extra debugging output:
```
mod_exists_regex = r'\s*/.+:\s*'
            for line in stderr.split('\n'):
                self.log.debug("Scanning line %s" % line)
                if OUTPUT_MATCHES['whitespace'].search(line):
                    self.log.debug("Output matches whitespace for module %s" % mod_name)
                    continue
                if OUTPUT_MATCHES['error'].search(line):
                    self.log.debug("Module show of module %s resulted in an error on the following line:" % mod_name)
                    self.log.debug("%s" % line)
                    break
                if re.match(mod_exists_regex, line):
                    self.log.debug("Module %s exists!)" % mod_name)
                    res = True
                break
            self.log.debug("New mod_exists_via_show will return %s" % res)
            return res
```
Resulted in:
```
== 2020-07-01 13:47:36,809 modules.py:566 DEBUG Scanning line module-version    Java/1.8.0_192 1.8
== 2020-07-01 13:47:36,810 modules.py:578 DEBUG New mod_exists_via_show will return False
```
In other words: none of the if-statements are true. The regexp doesn't match, even though I think it should (?).

**Comment by casparvl** [comment]
Ok, I think I found it. This line, returns the stderr of a module show command:
```
stderr = self.show(mod_name)
```
Looking at the debug logs, this runs

```
[casparl@int2 TMP]$ tclsh /hpc/eb/modules-tcl-1.775//libexec/modulecmd.tcl python show Java/1.8.0_192
```
of which the stderr output is
```
module-version  Java/1.8.0_192 1.8
-------------------------------------------------------------------
/home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/modulefiles/all/Java/1.8.0_192:

module-whatis   Description: Java Platform, Standard Edition (Java SE) lets you develop and deploy
 Java applications on desktops and servers.
module-whatis   Homepage: http://java.com/
module-whatis   URL: http://java.com/
conflict        Java
prepend-path    CMAKE_PREFIX_PATH       /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192
prepend-path    CPATH   /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/include
prepend-path    LD_LIBRARY_PATH /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/lib
prepend-path    LIBRARY_PATH    /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/lib
prepend-path    SURFSARA_LIBRARY_PATH   /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/lib
prepend-path    MANPATH /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/man
prepend-path    PATH    /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/bin
setenv          EBROOTJAVA      /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192
setenv          EBVERSIONJAVA   1.8.0_192
setenv          EBDEVELJAVA     /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192/easybuild/Java-1.8.0_192-easybuild-devel
prepend-path    PATH    /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192
setenv          JAVA_HOME       /home/casparl/.local/easybuild/RedHatEnterpriseServer7/2019/software/Java/1.8.0_192
-------------------------------------------------------------------
```
The new mod_exists_via_show function has this comment:
```
# - Check first non-whitespace line for something that looks like an absolute path terminated by a colon
```
That is indeed what the logic with the three if's and then a break does. This is also what causes my problem: for Tcl, the full path to the modulefile is on the third line, not the first!

Now the question is: how should one solve this. I'm happy to remove the `break` statement so that it goes on to the second and third line, but it was probably put in there to make the regexp more robust against matching other paths... At the very least it means we need to put the module name back in the regexp, because now it would match ANY path that ends in a colon with `r'\s*/.+:\s*'`.

**Comment by casparvl** [comment]
@Flamefire Since you wrote this piece of code, do you have a suggestion? Should we simply go back to the original regexp that was matched in EasyBuild <= 4.2.0? Or did you move away from that to solve a problem on your side?

**Comment by Flamefire** [comment]
> of which the stderr output is 

I'm confused by that. Are you 100% sure that the shown output is the stderr and not stdout + stderr? This is tested on CI so I'm surprised to see it failing.

If it is, then on to solving:
- Is there any chance to find out why tclsh is outputting the (I suppose) alias command from the modulerc and avoid that?
- the `---` lines are treated as whitespace, hence we need to check the second line

> I'm happy to remove the break statement so that it goes on to the second and third line, but it was probably put in there to make the regexp more robust against matching other paths...

I'd rather suggest to ignore the first line starting with `module-` to keep the strength

> At the very least it means we need to put the module name back in the regexp, because now it would match ANY path that ends in a colon with r'\s*/.+:\s*'.

This is by design. A module alias may alias a completely different name. E.g. imagine `Java` as an alias for `PrettySystemJRE` because of reasons. This is fine and should be supported.

So quick fix: Ignore the first `module-` line (maybe only on tclsh)

To strengthen it: Come up with a test for this. Again I thought this is already covered, maybe @boegel might know why it didn't come up on CI

**Comment by boegel** [comment]
@casparvl Which modules tool are you using here?

**Comment by boegel** [comment]
@Flamefire This issue didn't pop up in CI because it seems that @casparvl is using an ancient variant of a Tcl-based modules tool implementation we're not testing against:

```
$ module --version
Modules Release Tcl 1.923 (2017-07-20)
```

We are still testing against an old Tcl-based modules tool implementation in Travis (1.147), but the tests worked just fine there...

**Comment by boegel** [comment]
After some back and forth with @casparvl, here's the stderr output produced by `module show Java/1.8.0_192` that wasn't being digested well, because of using an old/anicent Tmod version + having multiple `.modulerc` files being picked up (which caused the warning):

```
module-version    Java/1.8.0_192 1.8
WARNING: Duplicate version symbol '1.8' found
module-version  Java/1.8.0_192 1.8
-------------------------------------------------------------------
/sw/arch/Debian10/EB_production/2019/modulefiles/lang/Java/1.8.0_192:
```

Fixing that in #3379
